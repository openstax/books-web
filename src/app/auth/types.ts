export interface User {
  firstName: string;
}

export interface State {
  user: User | undefined;
  established: boolean;
}
