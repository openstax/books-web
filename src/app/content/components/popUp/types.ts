import { LinkedArchiveTree, LinkedArchiveTreeNode, LinkedArchiveTreeSection } from '../../types';

export type LocationFilters = Map<
  string,
  { section: LinkedArchiveTreeNode, children?: LinkedArchiveTreeSection[] }
>;

export type LocationFiltersWithChildren = Map<
  string,
  { section: LinkedArchiveTree, children: LinkedArchiveTreeSection[] }
>;

export interface FiltersChange<T> {
  remove: T[];
  new: T[];
}
