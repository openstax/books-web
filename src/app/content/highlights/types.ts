import { Highlight, HighlightsSummary } from '@openstax/highlighter/dist/api';

export type HighlightData = Highlight;

export type CountsPerSource = Exclude<HighlightsSummary['countsPerSource'], undefined>;

export interface State {
  myHighlightsOpen: boolean;
  enabled: boolean;
  focused?: string;
  highlights: null | HighlightData[];
  summary: {
    filters: {
      colors: string[];
      chapters: string[];
    },
    pagination: {
      // even though we're manually splitting our requests out into smaller batches of sources,
      // if a source has more than [numPage] highlights the api will still paginate. in order to
      // access the subsequent pages, we need to store the original sources here, and continue to
      // use them until we run out of pages, then grab the next batch of sources. on the LAST
      // page the response size may be smaller than expected, in this case the next batch of
      // sources should be immediately queried to fill the remaining [numPage]. use
      // paginationUtils.getNextPageSources to get the next batch of sources
      sources: string[];
      page: number;
    } | null,
    loading: boolean;
    // totalCounts should reflect the UNFILTERED state of user's data, used in representing
    // available filter options. this should be updated on create/delete and after auth change
    totalCounts: {[key: string]: number};
    // filteredTotalCounts should reflect the FILTERED state of user's data, used in calculating
    // pagination requests. this should be updated after filter change before requesting
    // filtered highlights (because you need this to query them)
    filteredTotalCounts: {[key: string]: number};
    highlights: {
      [chapterId: string]: {[pageId: string]: HighlightData[]}
    };
  };
}
