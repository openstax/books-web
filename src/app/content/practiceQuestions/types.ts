import { LinkedArchiveTree, LinkedArchiveTreeSection } from '../types';

export interface State {
  isEnabled: boolean;
  summary: PracticeQuestionsSummary | null;
  open: boolean;
  selectedSection: LinkedArchiveTreeSection | null;
  currentQuestionIndex: number | null;
  questions: PracticeQuestion[];
}

export interface PracticeQuestionsSummary {
  countsPerSource: {
    [key: string]: number;
  };
}

interface PracticeAnswer {
  id: number;
  content_html: string;
  correctness: '0.0' | '1.0';
  feedback_html: string;
}

export interface PracticeQuestion {
  group_uuid: string;
  uid: string;
  stem_html: string;
  answers: PracticeAnswer[];
}

export type PracticeQuestions = PracticeQuestion[];

export type PracticeQuestionsLocationFilters = Map<
  string,
  { section: LinkedArchiveTree, children: LinkedArchiveTreeSection[] }
>;
