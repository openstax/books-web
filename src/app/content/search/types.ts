import { SearchResult } from '@openstax/open-search-client';

export interface State {
  results: SearchResult | null;
  loading: boolean;
  query: null | string;
}
