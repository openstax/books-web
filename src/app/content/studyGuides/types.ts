import { HighlightsSummary } from '@openstax/highlighter/dist/api';

export interface State {
  isEnabled: boolean;
  summary: HighlightsSummary | null;
}
