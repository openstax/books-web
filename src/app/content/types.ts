import { RouteParams, RouteState } from '../navigation/types';
import { content } from './routes';
import { State as SearchState } from './search/types';

export interface Params {
  book: string;
  page: string;
}

export interface State {
  tocOpen: boolean | null;
  params?: Params;
  loading: {
    book?: string;
    page?: string;
  };
  search: SearchState;
  book?: Book;
  page?: Page;
  references: PageReferenceMap[];
}

export interface PageReferenceMap extends PageReference {
  match: string;
}

export interface PageReference {
  state: RouteState<typeof content>;
  params: RouteParams<typeof content>;
}

export interface Book {
  id: string;
  shortId: string;
  title: string;
  theme: 'blue' | 'green' | 'gray' | 'yellow' | 'deep-green' | 'light-blue' | 'orange';
  tree: ArchiveTree;
  version: string;
  slug: string;
  license: {
    name: string;
    version: string;
  };
  publish_date: string;
  authors: Array<{
    value: {
      name: string;
    }
  }>;
}

export interface Page {
  id: string;
  shortId: string;
  title: string;
  version: string;
}

export interface ArchiveTreeSection {
  id: string;
  shortId: string;
  title: string;
}

export interface LinkedArchiveTree extends ArchiveTree {
  parent?: LinkedArchiveTree;
}

export interface LinkedArchiveTreeSection extends ArchiveTreeSection {
  parent: LinkedArchiveTree;
}

export interface ArchiveTree extends ArchiveTreeSection {
  contents: Array<ArchiveTree | ArchiveTreeSection>;
}

export interface ArchiveBook {
  id: string;
  shortId: string;
  title: string;
  tree: ArchiveTree;
  version: string;
  license: {
    name: string;
    version: string;
  };
}

export interface ArchivePage {
  id: string;
  shortId: string;
  content: string;
  version: string;
  title: string;
}

export type ArchiveContent = ArchivePage | ArchiveBook;
