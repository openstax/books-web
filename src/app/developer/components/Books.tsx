import React from 'react';
import { BOOKS } from '../../../config';
import { H3 } from '../../components/Typography';
import { StyledContentLink } from '../../content/components/ContentLink';
import { Book } from '../../content/types';
import { findDefaultBookPage, makeUnifiedBookLoader } from '../../content/utils';
import withServices from '../../context/Services';
import { AppServices } from '../../types';
import Panel from './Panel';

interface Props {
  services: AppServices;
}
interface State {
  books: Book[];
}

class Books extends React.Component<Props, State> {
  public state: State = {
    books: [],
  };

  public async componentDidMount() {
    const {archiveLoader, osWebLoader} = this.props.services;
    const bookLoader = makeUnifiedBookLoader(archiveLoader, osWebLoader);

    const books = await Promise.all(Object.entries(BOOKS).map(([bookId, {defaultVersion}]) =>
      bookLoader(bookId, defaultVersion)
    ));

    this.setState({books});
  }

  public render() {
    const {books} = this.state;

    return <Panel title='Books'>
      <ul>
        {books.map((book) => <li key={book.slug}>
          {this.renderBookLink(book)}
        </li>)}
      </ul>
    </Panel>;
  }

  private renderBookLink(book: Book) {
    const page = findDefaultBookPage(book);
    return <H3><StyledContentLink book={book} page={page}>{book.title}</StyledContentLink></H3>;
  }
}

export default withServices(Books);
