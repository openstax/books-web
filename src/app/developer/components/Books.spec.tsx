import React from 'react';
import renderer from 'react-test-renderer';
import mockArchiveLoader from '../../../test/mocks/archiveLoader';
import mockOSWebLoader from '../../../test/mocks/osWebLoader';
import * as Services from '../../context/Services';
import { AppServices } from '../../types';
import Books from './Books';

describe('Routes', () => {
  let archiveLoader: AppServices['archiveLoader'];
  let osWebLoader: AppServices['osWebLoader'];
  const services = {} as AppServices;

  beforeEach(() => {
    archiveLoader = mockArchiveLoader();
    osWebLoader = mockOSWebLoader();
    (services as any).archiveLoader = archiveLoader;
    (services as any).osWebLoader = osWebLoader;
  });

  it('matches snapshot', async() => {
    const component = renderer.create(<Services.Provider value={services}>
      <Books />
    </Services.Provider>);

    // defer promises...
    await new Promise((resolve) => setTimeout(resolve, 1));

    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
