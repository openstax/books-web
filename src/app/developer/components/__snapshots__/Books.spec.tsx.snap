// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Routes matches snapshot 1`] = `
<div>
  <h2>
    REX Books
  </h2>
  <div>
    <h2>
      Test Book 1
    </h2>
  </div>
</div>
`;
